ratchet_axum/src/lib.rs

```diff
@@ -0,0 +1,162 @@
+// Port of hyper_tunstenite for fastwebsockets.
+// https://github.com/de-vri-es/hyper-tungstenite-rs
+//
+// Copyright 2021, Maarten de Vries maarten@de-vri.es
+// BSD 2-Clause "Simplified" License
+//
+// Copyright 2023 Divy Srivastava <dj.srivastava23@gmail.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//todo missing docs
+
+#![deny(
+    // missing_docs,
+    missing_copy_implementations,
+    missing_debug_implementations,
+    trivial_numeric_casts,
+    unstable_features,
+    unused_must_use,
+    unused_mut,
+    unused_imports,
+    unused_import_braces
+)]
+
+use std::pin::Pin;
+use std::task::Context;
+use std::task::Poll;
+
+use axum_core::body::Body;
+use base64;
+use base64::engine::general_purpose::STANDARD;
+use base64::Engine;
+use eyre::Report;
+use http::HeaderMap;
+use hyper::Response;
+use hyper_util::rt::TokioIo;
+use pin_project::pin_project;
+use sha1::Digest;
+use sha1::Sha1;
+
+type Error = Report;
+
+#[derive(Debug)]
+pub struct IncomingUpgrade {
+    key: String,
+    headers: HeaderMap,
+    on_upgrade: hyper::upgrade::OnUpgrade,
+    pub permessage_deflate: bool,
+}
+
+impl IncomingUpgrade {
+    pub fn upgrade(self) -> Result<(Response<Body>, UpgradeFut), Error> {
```

Comment:
@SirCipher I am trying to implement the upgrade like this but having trouble because `ExtensionProvider` is not `Send`. I tried it with the below setup, and was able to get it to compile/run, but running into this error:
```
called `Result::unwrap()` on an `Err` value: Error { inner: Inner { kind: IO, source: Some(Kind(UnexpectedEof)) } }
```

wondering if you have any advice here

```
     pub fn upgrade_2<E, F, Fut>(self, f: F, provider: E) -> Response<Body>
    where
        F: FnOnce(UpgradedServer<TokioIo<hyper::upgrade::Upgraded>, E::Extension>) -> Fut
            + Send
            + 'static,
        Fut: Future<Output = ()> + Send,
        E: ExtensionProvider + Send + 'static,
        <E as ExtensionProvider>::Extension: Send,
    {
        tokio::spawn(async move {
            let upgraded = match self.on_upgrade.await {
                Ok(upgraded) => upgraded,
                Err(err) => {
                    return;
                }
            };
            let upgraded = TokioIo::new(upgraded);
            let upgrade_server = accept_with(
                upgraded,
                WebSocketConfig::default(),
                provider,
                SubprotocolRegistry::default(),
            )
            .await
            .unwrap()
            .upgrade_with(self.headers)
            .await
            .unwrap();

            f(upgrade_server).await;
        });
        let builder = Response::builder()
            .status(hyper::StatusCode::SWITCHING_PROTOCOLS)
            .header(hyper::header::CONNECTION, HEADER_CONNECTION)
            .header(hyper::header::UPGRADE, HEADER_UPGRADE)
            .header(hyper::header::SEC_WEBSOCKET_ACCEPT, self.key);

        let response = builder
            .body(Body::default())
            .expect("bug: failed to build response");
        response
    }
```